core/container.go

```diff
@@ -630,6 +645,30 @@ func (container *Container) withMounted(target string, srcDef *pb.Definition, sr
 	return &Container{ID: id}, nil
 }
 
+func (container *Container) updateRootFS(ctx context.Context, gw bkgw.Client, subdir string, fn func(dir *Directory) (*Directory, error)) (*Container, error) {
+	dir, mount, err := locatePath(ctx, container, subdir, gw, NewDirectory)
+	if err != nil {
+		return nil, err
+	}
+
+	dir, err = fn(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	// If not in a mount, replace rootfs
+	if mount == nil {
+		return container.WithFS(ctx, dir)
+	}
+
+	dirPayload, err := dir.ID.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	return container.withMounted(mount.Target, dirPayload.LLB, "/")
```

Comment:
Hey, thanks for this!

I actually am not sure what I'm breaking vs what's already broken. I opened this earlier: https://github.com/dagger/dagger/issues/3903

There's something not right with path chaining in the API

Reply:
@vito May I have some help with this PR when you have some time?

Reply:
Taking a look!

Reply:
This seems to fix it:

```suggestion
	return container.withMounted(mount.Target, dirPayload.LLB, mount.SourcePath)
```

> It should be the subselection of the original dir rather than / (I think...)

The `dirPayload.Dir` being mounted is `/a/b/foo`; `locatePath` makes it relative to the mount's SourcePath so we just need to continue respecting it with the new mount:

https://github.com/dagger/dagger/blob/eb9ef56fc4958bf5e1ac781533cdf0cb92da4401/core/container.go#L577-L584

Admittedly this is all pretty confusing coming back after a week of vacation, but the fix at least _looks_ straightforward. :stuck_out_tongue: 

Should definitely have a test for this though.

Here's one:

```go
func TestContainerWithDirectoryToMount(t *testing.T) {
	t.Parallel()

	ctx := context.Background()

	c, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stdout))
	require.NoError(t, err)
	defer c.Close()

	mnt := c.Directory().
		WithNewDirectory("/top/sub-dir/sub-file").
		Directory("/top") // <-- the important part!
	ctr := c.Container().
		From("alpine:3.16.2").
		WithMountedDirectory("/mnt", mnt)

	dir := c.Directory().
		WithNewFile("/copied-file", dagger.DirectoryWithNewFileOpts{
			Contents: "some-content",
		})

	ctr = ctr.WithDirectory("/mnt/sub-dir/copied-dir", dir)

	contents, err := ctr.Exec(dagger.ContainerExecOpts{
		Args: []string{"find", "/mnt"},
	}).Stdout().Contents(ctx)
	require.NoError(t, err)
	require.Equal(t, []string{
		"/mnt",
		"/mnt/sub-dir",
		"/mnt/sub-dir/sub-file",
		"/mnt/sub-dir/copied-dir",
		"/mnt/sub-dir/copied-dir/copied-file",
	}, strings.Split(strings.Trim(contents, "\n"), "\n"))
}
```

Without the fix, it fails like so:

```
                Error:          Not equal: 
                                expected: []string{"/mnt", "/mnt/sub-dir", "/mnt/sub-dir/sub-file", "/mnt/sub-dir/copied-dir", "/mnt/sub-dir/copied-dir/copied-file"}
                                actual  : []string{"/mnt", "/mnt/top", "/mnt/top/sub-dir", "/mnt/top/sub-dir/sub-file", "/mnt/top/sub-dir/copied-dir", "/mnt/top/sub-dir/copied-dir/copied-fi
le"}
```